### PR TITLE
Fixing allocation domain for index_select

### LINKED ISSUE #1122: index_select + allocationDomain breaks index_compute: `it != p2c.end() INTERNAL ASSERT FAILED`
This is the error I run into:
```
RuntimeError: it != p2c.end() INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/index_compute.cpp":59, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues.
Exception raised from getProducerHaloOffset at /opt/pytorch/nvfuser/csrc/index_compute.cpp:59 (most recent call first):
frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, char const*) + 0x8f (0x7fdbef8c64fc in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #1: nvfuser::nvfCheckMsgImpl(char const*, char const*) + 0 (0x7fdbf0bb30b4 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #2: <unknown function> + 0x14022bf (0x7fdbef9cc2bf in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #3: <unknown function> + 0x14024d0 (0x7fdbef9cc4d0 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #4: nvfuser::Index::getProducerAllocationIndices(nvfuser::TensorView*, nvfuser::TensorView const*, std::vector<nvfuser::kir::ForLoop*, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_set<nv
fuser::kir::ForLoop*, std::hash<nvfuser::kir::ForLoop*>, std::equal_to<nvfuser::kir::ForLoop*>, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_map<nvfuser::IterDomain*, nvfuser::Val*, std::hash<
nvfuser::IterDomain*>, std::equal_to<nvfuser::IterDomain*>, std::allocator<std::pair<nvfuser::IterDomain* const, nvfuser::Val*> > > const&) + 0x9ce (0x7fdbef9d7e76 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_
codegen.so)
frame #5: nvfuser::Index::getGlobalProducerStridedIndices(nvfuser::TensorView*, nvfuser::TensorView const*, std::vector<nvfuser::kir::ForLoop*, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_set
<nvfuser::kir::ForLoop*, std::hash<nvfuser::kir::ForLoop*>, std::equal_to<nvfuser::kir::ForLoop*>, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_map<nvfuser::IterDomain*, nvfuser::Val*, std::ha
sh<nvfuser::IterDomain*>, std::equal_to<nvfuser::IterDomain*>, std::allocator<std::pair<nvfuser::IterDomain* const, nvfuser::Val*> > > const&) + 0x9a (0x7fdbef9d43bc in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #6: nvfuser::Index::getProducerStridedIndices(nvfuser::TensorView*, nvfuser::TensorView const*, std::vector<nvfuser::kir::ForLoop*, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_set<nvfuser::kir::ForLoop*, std::hash<nvfuser::kir::ForLoop*>, std::equal_to<nvfuser::kir::ForLoop*>, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_map<nvfuser::IterDomain*, nvfuser::Val*, std::hash<nvfuser::IterDomain*>, std::equal_to<nvfuser::IterDomain*>, std::allocator<std::pair<nvfuser::IterDomain* const, nvfuser::Val*> > > const&, bool) + 0x119 (0x7fdbef9d9a57 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #7: nvfuser::Index::getProducerIndex(nvfuser::TensorView*, nvfuser::TensorView const*, std::vector<nvfuser::kir::ForLoop*, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_set<nvfuser::kir::ForLoop*, std::hash<nvfuser::kir::ForLoop*>, std::equal_to<nvfuser::kir::ForLoop*>, std::allocator<nvfuser::kir::ForLoop*> > const&, std::unordered_map<nvfuser::IterDomain*, nvfuser::Val*, std::hash<nvfuser::IterDomain*>, std::equal_to<nvfuser::IterDomain*>, std::allocator<std::pair<nvfuser::IterDomain* const, nvfuser::Val*> > > const&, bool) + 0x5c (0x7fdbef9d9d62 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #8: nvfuser::IndexLowering::lowerSrcIndex(nvfuser::Val*, nvfuser::Val*, std::unordered_map<nvfuser::IterDomain*, nvfuser::Val*, std::hash<nvfuser::IterDomain*>, std::equal_to<nvfuser::IterDomain*>, std::allocator<std::pair<nvfuser::IterDomain* const, nvfuser::Val*> > > const&, bool) const + 0xe8 (0x7fdbefc1377a in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #9: nvfuser::IndexLowering::handle(nvfuser::IndexSelectOp const*) + 0x228 (0x7fdbefc16022 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
frame #10: void nvfuser::Expr::constDispatch<nvfuser::OptOutConstDispatch*>(nvfuser::OptOutConstDispatch*, nvfuser::Expr const*) + 0x426 (0x7fdbef7f64c2 in /opt/pytorch/nvfuser/nvfuser/lib/libnvfuser_codegen.so)
```

Repro scripts:
```
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id4183(fd : FusionDefinition) -> None :
    T1 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=DataType.Int, is_cpu=False, stride_order=[0, 1])
    #T1 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=DataType.Int, is_cpu=False)
    S2 = fd.define_scalar(4, dtype=DataType.Int)
    S3 = fd.define_scalar(0, dtype=DataType.Int)
    S4 = fd.define_scalar(8, dtype=DataType.Int)
    T5 = fd.ops.iota(S2, S3, S4, dtype=DataType.Int)
    #fd.add_output(T5)
    T25 = fd.ops.index_select(T1, T5, dim=1)
    fd.add_output(T25)

with FusionDefinition() as fd:
    nvfuser_fusion_id4183(fd)

inputs = [
    torch.randint(0, 10, (252,), dtype=torch.int64, device='cuda:0').as_strided((9, 28), (1, 9)),
    #torch.randint(0, 10, (252,), dtype=torch.int64, device='cuda:0').as_strided((9, 28), (28, 1)),
]
o = fd.execute(inputs)
```

Code changes here does fix the issue: #1113, however, it simultaneously breaks a bunch of existing AllocationDomainTest
```
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] AllocationDomainTest.NHWC1d_To_NHWC4d
[  FAILED  ] AllocationDomainTest.NHWC1d_To_NHWC1d
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d_cacheBefore
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d_cacheAfter
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d_cacheFork
```

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1122 
1. switch to use allocation domain instead of rfactor domain in `getProducerHaloOffset` for better consistency;
2. relax the map check. Since some existing tests happen to rely on those.
3. adding a test case for index_select with permuted inputs.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
Nope... this didn't work.... something else went wrong.

**Comment by jjsjann123** [comment]
```
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] AllocationDomainTest.NHWC1d_To_NHWC4d
[  FAILED  ] AllocationDomainTest.NHWC1d_To_NHWC1d
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d_cacheBefore
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d_cacheAfter
[  FAILED  ] AllocationDomainTest.NHWC2d_To_NHWC2d_cacheFork
```

File: csrc/index_compute.cpp [hunk 1]
```diff
@@ -48,15 +48,19 @@ int getProducerHaloOffset(
                  .mapDifferentExtents(true)
                  .mapProducerToConsumer();
 
-  auto producer_id = producer_tv->getMaybeRFactorDomain()[producer_axis];
+  auto producer_id = producer_tv->getMaybeAllocationDomain()[producer_axis];
```
Comment (jjsjann123) [inline csrc/index_compute.cpp:51]: This change is just so we have better consistency with our usage of `producer_axis`

**Comment by jjsjann123** [comment]
!build

File: csrc/index_compute.cpp [hunk 2]
```diff
@@ -48,15 +48,19 @@ int getProducerHaloOffset(
...
   // where no mapping exists for a producer axis is when it is a
   // reduction axis. Since this function is only used for indexing
   // producer tensors, where reduction axes are skipped, producer_id
   // should never be a reduction axis.
-  NVF_ERROR(it != p2c.end());
+  if (it == p2c.end()) {
+    TORCH_WARN(
+        "getProducerHaloOffset p2c mapping has failed. See "
+        "https://github.com/NVIDIA/Fuser/issues/1122");
+  }
```
Comment (zasdfgbnm) [inline csrc/index_compute.cpp:63]: Should we return 0 here? The next line is illegal memory access.

Comment (jjsjann123) [reply csrc/index_compute.cpp:63]: err. good catch.. didn't copy the whole thing over. :face_exhaling:

**Comment by jjsjann123** [comment]
!build

**Review by zasdfgbnm — APPROVED** [review]
This is bad, but not worse than without this change. Approve to unblock.

**Pull request merged**